### PR TITLE
Enable gpu compat memory on for output allocation dataset ops

### PR DESCRIPTION
**Pull request by kushanam**
This MR is a follow up to [55594](https://github.com/tensorflow/tensorflow/pull/55594), and the potential merge should happen afterwards. The goal is to activate the use_gpu_compat_allocator attribute to allocate gpu compatible memory directly prior to the prefetch_to_device op. The list of affected ops are:
RepeatDataset, MapDataset, ParallelMapDataset, ParallelMapDatasetV2, InterleaveDataset, ParallelInterleaveDatasetV2, ParallelInterleaveDatasetV3, ParallelInterleaveDatasetV4, FilterDataset, ParallelFilterDataset, BatchDataset, BatchDatasetV2, ParallelBatchDataset, ShardDataset, PaddedBatchDataset, PaddedBatchDatasetV2, RangeDataset
Attn: @changhuilin

### CODE REVIEW

**Comment by gbaned** [comment]
@kushanam Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
@kushanam Can you please check @qqfish's comments and resolve conflicts?. Thank you!

**Comment by gbaned** [comment]
@kushanam Any update on this PR? Please. Thank you!

File: tensorflow/core/framework/dataset.h [hunk 1]
```diff
@@ -1336,6 +1337,10 @@ class DatasetOpKernel : public OpKernel {
                   errors::InvalidArgument(absl::StrCat(
                       "Could not parse the 'metadata' attribute.")));
     }
+    if (ctx->HasAttr(UseGpuCompatAllocatorAttr))
+    {
```
[the remark this answers is not in the document]
Comment (kushanam) [reply tensorflow/core/framework/dataset.h:1341]: fixed.

File: tensorflow/core/framework/dataset.cc [hunk 2]
```diff
@@ -997,7 +997,16 @@ void DatasetOpKernel::Compute(OpKernelContext* ctx) {
   MakeDataset(ctx, &dataset);
   if (ctx->status().ok()) {
     Tensor* output = nullptr;
-    OP_REQUIRES_OK(ctx, ctx->allocate_output(0, TensorShape({}), &output));
+    if (UseGpuCompatAllocatorAttr_)
+    {
```
[the remark this answers is not in the document]
Comment (kushanam) [reply tensorflow/core/framework/dataset.cc:1001]: Moved.

File: tensorflow/core/framework/dataset.cc [hunk 3]
```diff
@@ -997,7 +997,16 @@ void DatasetOpKernel::Compute(OpKernelContext* ctx) {
...
+      AllocatorAttributes attr;
+      attr.set_gpu_compatible(true);
+      OP_REQUIRES_OK(ctx, ctx->allocate_output(0, TensorShape({}), &output, attr));
+    }
+    else
```
[the remark this answers is not in the document]
Comment (kushanam) [reply tensorflow/core/framework/dataset.cc]: Done.

File: tensorflow/core/framework/dataset.cc [hunk 4]
```diff
@@ -997,7 +997,16 @@ void DatasetOpKernel::Compute(OpKernelContext* ctx) {
...
+    {
```
[the remark this answers is not in the document]
Comment (kushanam) [reply tensorflow/core/framework/dataset.cc]: Moved.

**Comment by kushanam** [comment]
> @kushanam Any update on this PR? Please. Thank you!

Requested changes are applied. Thanks.

**Pull request closed without merging**